### ISSUE TITLE
Fix broken tests in test_breaking_into_raised_and_unhandled_exceptions

### ISSUE THREAD

**Issue by DonJayamanne** [Microsoft/ptvsd#723]
_From @DonJayamanne on August 1, 2018 5:25_

Currently skipped

_Copied from original issue: Microsoft/vscode-python#2293_

## Repro steps
1. Remove the line skipping this test in `test_exception.py`
2. `set PYTHONPATH=C:\ptvsd`
3. Run test:
```
py -3.7 -m tests -v tests.system_tests.test_exceptions.PTVSDAttachModuleExceptionLifecycleTests.test_breaking_into_raised_and_unhandled_exceptions
```
4. Should pass.

The test currently fails because of the unexpected third thread stopped event.

**Comment by karthiknadig** [comment]
The cause for this issue is same as #679.

**Comment by karthiknadig** [comment]
Looks like there is another issue, exposed after fixing 679.
@fabioz. looks like we get uncaught exception event raised twice in some cases. These are the call stacks for one such case. see above for repro steps.
**First event: on raised (this seems correct)**
```
File "C:\\Python37\\lib\\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
File "C:\\Python37\\lib\\runpy.py", line 85, in _run_code
    exec(code, run_globals)
File "C:\\GIT\\ptvsd\\tests\\resources\\system_tests\\test_exceptions\\mypkg_attach_unhandled\\__main__.py", line 12, in <module>
    main()
File "C:\\GIT\\ptvsd\\tests\\resources\\system_tests\\test_exceptions\\mypkg_attach_unhandled\\__main__.py", line 8, in main
    raise ArithmeticError(\'Hello\')
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 146, in trace_exception
    self.handle_exception(frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 334, in handle_exception
    self.do_wait_suspend(thread, frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 134, in do_wait_suspend
    self._args[0].do_wait_suspend(*args, **kwargs)
```

**Second event: uncaught (i suspect is the wrong one)**
```
File "C:\\Python37\\lib\\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
File "C:\\Python37\\lib\\runpy.py", line 85, in _run_code
    exec(code, run_globals)
File "C:\\GIT\\ptvsd\\tests\\resources\\system_tests\\test_exceptions\\mypkg_attach_unhandled\\__main__.py", line 12, in <module>
    main()
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_trace_dispatch_regular.py", line 121, in trace_dispatch
    return thread_tracer.__call__(frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_trace_dispatch_regular.py", line 316, in __call__
    ).trace_dispatch(frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 446, in trace_dispatch
    self.handle_exception(frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 334, in handle_exception
    self.do_wait_suspend(thread, frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_frame.py", line 134, in do_wait_suspend
    self._args[0].do_wait_suspend(*args, **kwargs)
```

**Third event: uncaught (this seems correct)**
```
File "C:\\Python37\\lib\\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_trace_dispatch_regular.py", line 199, in trace_dispatch_and_unhandled_exceptions
    self.trace_unhandled_exceptions(frame, event, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_trace_dispatch_regular.py", line 189, in trace_unhandled_exceptions
    stop_on_unhandled_exception(py_db, t, additional_info, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_breakpoints.py", line 156, in stop_on_unhandled_exception
    py_db.stop_on_unhandled_exception(thread, frame, frames_byid, arg)
File "C:\\GIT\\ptvsd\\ptvsd\\_vendored\\pydevd\\pydevd.py", line 893, in stop_on_unhandled_exception
    self.do_wait_suspend(thread, frame, \'exception\', arg, "trace")
```

**Comment by DonJayamanne** [comment]
@karthiknadig
How did we catch this issue you have reported above?
Via CI tests or your general local testing?
If it was when performing local testing, please could we create an issue to create a test that reproduces this bug to ensure it gets caught again in CI.

**Comment by karthiknadig** [comment]
I removed the skip for all tests in `test_exceptions.py`, and ran them. This tests will catch the issue:
`PTVSDAttachModuleExceptionLifecycleTests.test_breaking_into_raised_and_unhandled_exceptions`

**Comment by fabioz** [comment]
This appears correct to me (twice in caught exceptions and once in unhandled exceptions).

i.e.:
- The case for hitting twice when handled unhandled is also reported at: https://github.com/Microsoft/ptvsd/issues/653 (we had agreed it seemed reasonable, so, I'm considering it's Ok).

- Caught exceptions are usually reported once when it's raised and another in each caller until it's handled or breaks the program execution (so, in the given scenario it should hit twice), although note but there's a provision in `pydevd` to notify about it only once depending on the parameters passed to `CMD_ADD_EXCEPTION_BREAK` -- see:
https://github.com/fabioz/PyDev.Debugger/blob/master/_pydevd_bundle/pydevd_process_net_command.py#L557 for comments on the options there (so, it seems ok in `pydevd`, although `ptvsd` may want to make that configurable or change the default).

**Comment by karthiknadig** [comment]
For now we are keeping the existing behavior.
